pkg/controller/endpoint/trigger_time_tracker.go

```diff
@@ -0,0 +1,147 @@
+// shouldn't be called concurrently for the same key! This contract is fulfilled in the current
+// implementation of the endpoints controller.
+func (t *TriggerTimeTracker) ComputeEndpointsLastChangeTriggerTime(
+	key string, service *v1.Service, pods []*v1.Pod) time.Time {
+
+	// As there won't be any concurrent calls for the same key, we need to guard access only to the
+	// endpointsStates map.
+	t.mutex.Lock()
+	state, wasKnown := t.endpointsStates[key]
+	t.mutex.Unlock()
+
+	// Update the state before returning.
+	defer func() {
+		t.mutex.Lock()
+		t.endpointsStates[key] = state
+		t.mutex.Unlock()
+	}()
+
+	var minChangedTriggerTime time.Time
+	// TODO(mm4tt): If memory allocation / GC performance impact of recreating map in every call
+	// turns out to be too expensive, we should consider rewriting this to reuse the existing map.
+	podTriggerTimes := make(map[string]time.Time)
+	for _, pod := range pods {
+		if podTriggerTime := getPodTriggerTime(pod); !podTriggerTime.IsZero() {
+			podTriggerTimes[pod.Name] = podTriggerTime
+			if podTriggerTime.After(state.lastPodTriggerTimes[pod.Name]) {
+				// Pod trigger time has changed since the last sync, update minChangedTriggerTime.
+				minChangedTriggerTime = min(minChangedTriggerTime, podTriggerTime)
+			}
+		}
+	}
+	serviceTriggerTime := getServiceTriggerTime(service)
+	if serviceTriggerTime.After(state.lastServiceTriggerTime) {
+		// Service trigger time has changed since the last sync, update minChangedTriggerTime.
+		minChangedTriggerTime = min(minChangedTriggerTime, serviceTriggerTime)
+	}
+
+	state.lastPodTriggerTimes = podTriggerTimes
+	state.lastServiceTriggerTime = serviceTriggerTime
+
+	if !wasKnown {
+		// New Endpoints object / new Service, use Service creationTimestamp.
+		return service.CreationTimestamp.Time
+	} else {
+		// Regular update of the Endpoints object, return min of changed trigger times.
+		return minChangedTriggerTime
+	}
+}
+
+// DeleteEndpoints deletes endpoints state stored in this util.
+func (t *TriggerTimeTracker) DeleteEndpoints(key string) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	delete(t.endpointsStates, key)
+}
+
+// getPodTriggerTime returns the time of the pod change (trigger) that resulted or will result in
+// the endpoints object change.
+func getPodTriggerTime(pod *v1.Pod) (triggerTime time.Time) {
+	if readyCondition := podutil.GetPodReadyCondition(pod.Status); readyCondition != nil {
+		triggerTime = readyCondition.LastTransitionTime.Time
+	}
+	// TODO(mm4tt): Implement missing cases: deletionTime set, pod label change
+	return triggerTime
+}
+
+// getServiceTriggerTime returns the time of the service change (trigger) that resulted or will
+// result in the endpoints object change.
+func getServiceTriggerTime(service *v1.Service) (triggerTime time.Time) {
+	// TODO(mm4tt): Ideally we should look at service.LastUpdateTime, but such thing doesn't exist.
```

Comment:
Ignoring service update may actually result in discrepancy right? 
For example:
Service ports got updated and no change on pods. This would result in update in endpoints object. The `minChangedTriggerTime` will still return one of the time delta after the latest pod update, right?

In this case, unless we have the something similar to service.LastUpdateTime, I think we can ignore it and skip recording. WDYT?

Reply:
That's what we do. The minChangedTriggerTime definition is: the min trigger time of all trigger times that **have changed** since the last sync.

Assuming that only service has changed, the minChangeTriggerTime would be a 'zero' time, as nothing has changed (since we don't have the service.LastUpdateTime and service.CreationTime stays the same). If the ComputeEndpointsLastChangeTriggerTime returns a 'zero' time the result should be ignored and the EndpointsLastChangeTriggerTime shouldn't be exported.

If in future we manage to get the service.LastUpdateTime somehow it will be enough to just change this place and everything will work correctly. That's why I left a TODO here.

I updated the docs and comments to make it easier to understand. Also added a test for that case.